torch/distributed/fsdp/_optim_utils.py

```diff
@@ -1323,3 +1372,100 @@ def return_fn(fqn_to_param_info):
         return_fn,
         fqn_to_param_info,
     )
+
+
+def _gather_orig_param_state(
+    fqn: str,
+    fsdp_param_info: FSDPParamInfo,
+    optim_state: Dict[str, Any],
+    shard_state: bool,
+) -> Dict[str, Any]:
+    """
+    Gather the optimizer state for the original parameter with the name ``fqn``.
+    This API should only be used when ``use_orig_params`` is True.
+    """
+    fsdp_state = fsdp_param_info.state
+    flat_param = fsdp_param_info.flat_param
+    param_idx = fsdp_param_info.fqn_indices[fqn]
+    if (
+        fsdp_state.world_size == 1
+        or fsdp_state.sharding_strategy == ShardingStrategy.NO_SHARD
+    ):
+        return optim_state
+
+    # Gathering state from all ranks. This step may be slow. However,
+    # `state_dict()` is not in the critical path. We can fuse the communication
+    # if the performance becomes a problem.
+    state_objects = {
+        state_name: value for state_name, value in sorted_items(optim_state)
+    }
+    object_list: List[Dict[str, Any]] = [
+        {} for _ in range(cast(int, fsdp_state.world_size))
+    ]
+    dist.all_gather_object(object_list, state_objects)
+    orig_state: Dict[str, Any] = {}
+    for state in object_list:
+        for state_name, value in state.items():
+            curr_value = orig_state.get(state_name, [])
+            if torch.is_tensor(value):
+                if value.dim() > 0:
+                    curr_value.append(value)
+                elif torch.is_tensor(curr_value):
+                    assert torch.equal(curr_value, value)
+                else:
+                    orig_state[state_name] = value
+            else:
+                assert curr_value == [] or curr_value == value
+                orig_state[state_name] = value
+
+    # Unflatten state values.
+    for state_name in orig_state.keys():
+        value = orig_state[state_name]
+        if not isinstance(value, list) or not torch.is_tensor(value[0]):
+            continue
+        value = torch.concat(value)[: flat_param._numels[param_idx]].reshape(
```

Comment:
Conceptual: In what case would `torch.concat(value)` have more numel than `flat_param._numels[param_idx]` (meaning that we are truncating)?

Reply:
Excellent point. The padding shouldn't be visible here. However, I'm not sure if it is always true as we were discussing the cost of F.pad and may change the implementation. This is just a safety guard. If you believe padding will never be seen, I can remove it.

Reply:
Ah good catch. Let us keep the trimming logic here. The safe guard is good.